### PR TITLE
Chitter - pinky theme

### PR DESCRIPTION
**Pull request by cspoppuppy**
Functions
Sign up. Prevent register the same email or username twice, with error message in red.
Sign in. Reject wrong email or password, with error message in red.
View messages without the need to sign in. Messages are listed from latest to oldest.
Can post message after sign in.
Can send email to tagged user

Overview
Covered all functions, test coverage 100%
Responsive web pages (scale up/down when resize browser window)
Styling
Deal with apostrophes in text fields before passing to database

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,135 +1,54 @@
...
-So that I can better appreciate the context of a peep
-I want to see the time at which it was made
+1. [Setup](docs/setup.md)
+2. Run command `rackup`
+3. In web browser, enter i.e. "localhost:9292" (or the port provided from previous step)
+4. Sign up a few users (some with your real email addresses, best with gmail), play around to register same email and username twice to see the error messages
+5. Sign out
+6. Sign in with registered users, play around with wrong email or password
+7. Chit a few messages, latest will appear at the top
+8. Chit messages with @username (replace username with real usernames registered), choose the ones with real email addresses. You should receive email notifications (check junk if not)
```
Comment (inbarvernia) [inline README.md:28]: Really great instructions on how to set-up and interact!

File: README.md [hunk 2]
```diff
@@ -1,135 +1,54 @@
...
 
-As a Maker
-So that I can post messages on Chitter as me
-I want to sign up for Chitter
+*Tried with hotmail, seems not getting the notification. With gmail, getting it straightaway in junk folder.*
```
Comment (inbarvernia) [inline README.md:30]: I've tried signing-up using Gmail, but for some reason did not get an e-mail - not in junk mail folder either.

File: README.md [hunk 3]
```diff
@@ -1,135 +1,54 @@
...
-So that I can avoid others posting messages on Chitter as me
-I want to log out of Chitter
+---------
+### Overview
 
-ADVANCED
+- Covered all functions, test coverage 100%
+- Responsive web pages (scale up/down when resize browser window)
+- Styling
+- Deal with apostrophes in text fields before passing to database
```
Comment (inbarvernia) [inline README.md:45]: 👏

File: db/migrations/create_db_tables.sql [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  password VARCHAR(140),
+  name VARCHAR(200),
+  username VARCHAR(100) UNIQUE
+);
+
+CREATE TABLE messages (
+  id SERIAL PRIMARY KEY,
+  content VARCHAR(2000),
+  user_id INTEGER REFERENCES users(id) ON DELETE CASCADE,
+  created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP::TIMESTAMP(0)
```
Comment (inbarvernia) [inline db/migrations/create_db_tables.sql:13]: Great to know there's a TIMESTAMP data type in SQL! I had assumed you're have to use Ruby's Time class to create this information.

File: db/migrations/create_db_tables.sql [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+);
+
+CREATE OR REPLACE VIEW view_messages AS 
+  SELECT m.*, u.name, u.username
+  FROM messages m
+  LEFT JOIN users u
+  ON m.user_id = u.id;
```
Comment (inbarvernia) [inline db/migrations/create_db_tables.sql:20]: Interesting - so, you are formatting the message information directly here, so that it's easier to work with directly in Ruby in the view later?

File: public/css/style.css [hunk 6]
```diff
@@ -0,0 +1,119 @@
+body {
```
Comment (inbarvernia) [inline public/css/style.css:1]: I really liked your design and colour scheme!

File: spec/features/registration_spec.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  scenario 'a user can not sign up with same email twice' do
+    visit '/users/new'
+    fill_in('email', with: 'test@example.com')
+    fill_in('password', with: 'password123')
+    fill_in('name', with: 'testname2')
+    fill_in('username', with: 'testusername2')
+    click_button('Submit')
+
+    expect(page).to_not have_content "Welcome, testname2"
```
Comment (inbarvernia) [inline spec/features/registration_spec.rb:21]: It would be nice here and in following feature test to also expect page to have the appropriate error message, e.g. 'Username or e-mail already in use" - I see these errors are already in place and working in the actual app, but it might be good to include them in the tests as well.

File: spec/unit/message_spec.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      user = User.create('test7@example.com', 'password123', 'test7', 'test7')
+      message = Message.add('chit chit', user.id)
+      expect(message.content).to eq 'chit chit'
+      expect(message.username).to eq 'test7'
+    end
+  end
+
+  describe '.all' do
+    it 'call for all messages' do
+      expect(Message.all.first).to be_a Message
```
Comment (inbarvernia) [inline spec/unit/message_spec.rb:16]: Possibly could do something like Message.all.each here instead of .first, to make sure every item in all is a Message instance?

File: spec/unit/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    it 'hashes the password using BCrypt' do
+      expect(BCrypt::Password).to receive(:create).with('password123')
+      User.create('test1@example.com', 'password123', 'test1', 'test1')
+    end
+
+    it 'creates a new user' do
+      user = User.create('test2@example.com', 'password123', 'test2', 'test2')
+      # persisted_data = persisted_data(user, user.id)
+      expect(user).to be_a User
+      # expect(user.id).to eq persisted_data.first['id']
```
Comment (inbarvernia) [inline spec/unit/user_spec.rb:14]: You can probably delete the commented-out lines 12 and 14 - I'm guessing they are left over from an earlier test. :)

**Review by inbarvernia — COMMENTED** [review]
Such a great project, Sha! :) Very clean and clear code - you've used some things I haven't used before, but it was so clear that I could still understand what it was doing even without being familiar with it. Lots of good thinking of edge cases and "unhappy paths" built in, and also great ideas for additional ways to develop it further. Very good, thorough testing, and feature tests are also very clear, so that even if I hadn't set-up the app on my computer, it would have been easy to understand how it should work. Lovely visual style as well! Really inspiring. :)

File: spec/features/registration_spec.rb (hunk 7 above)
Comment (cspoppuppy) [reply spec/features/registration_spec.rb:21]: good point!

File: spec/unit/message_spec.rb (hunk 8 above)
Comment (cspoppuppy) [reply spec/unit/message_spec.rb:16]: another good point~

File: spec/unit/user_spec.rb (hunk 9 above)
Comment (cspoppuppy) [reply spec/unit/user_spec.rb:14]: yes, forgot to remove 😂

File: db/migrations/create_db_tables.sql (hunk 5 above)
Comment (cspoppuppy) [reply db/migrations/create_db_tables.sql:20]: Not sure what's the best practice in doing so, did it because it is the easiest way I can come up at the moment

**Review by penguat — COMMENTED** [review]
Well done Sha, this looks really good to me!

I've not addressed Travis failing inline, but you asked me to look at that as well. I believe, based on your screenshot, that Travis failed to find one of your packages, and that's why it failed. 
Normally something being deprecated is a warning, indicating that you should stop using it and use an alternative approach, as whatever's deprecated will be removed in a future release - that means it should still work for now.
